mapproxy/cache/geopackage.py

```diff
@@ -0,0 +1,657 @@
+
+    def cleanup(self):
+        """
+        Close all open connection and remove them from cache.
+        """
+        if getattr(self._db_conn_cache, 'db', None):
+            self._db_conn_cache.db.close()
+        self._db_conn_cache.db = None
+
+    @staticmethod
+    def _check_table_name(table_name):
+        """
+        >>> GeopackageCache._check_table_name("test")
+        'test'
+        >>> GeopackageCache._check_table_name("test_2")
+        'test_2'
+        >>> GeopackageCache._check_table_name("test3;")
+        Traceback (most recent call last):
+        ...
+        ValueError: The table_name test3; contains unsupported characters.
+        >>> GeopackageCache._check_table_name("table name")
+        Traceback (most recent call last):
+        ...
+        ValueError: The table_name table name contains unsupported characters.
+
+        @param table_name: A desired name for an geopackage table table.
+        @return: The name of the table if it is good, other wise and exception.
+        """
+
+        if is_alnum(table_name):
+            return table_name
+        else:
+            log.info("The table name may only contain alphanumeric characters or an underscore.")
+            raise ValueError("The table_name {0} contains unsupported characters.".format(table_name))
+
+    def ensure_gpkg(self):
+        if not os.path.isfile(self.geopackage_file):
+            with FileLock(os.path.join(os.path.dirname(self.geopackage_file), 'init.lck'),
+                          remove_on_unlock=True):
+                ensure_directory(self.geopackage_file)
+                self._initialize_gpkg()
+        else:
+            if not self.check_gpkg():
+                ensure_directory(self.geopackage_file)
+                self._initialize_gpkg()
+
+    def check_gpkg(self):
+        if not self._verify_table():
+            print("table invalid")
+            return False
+        if not self._verify_gpkg_contents():
+            print("gpkg_contents invalid")
+            return False
+        if not self._verify_tile_size():
+            print("tile_size invalid")
+            return False
+        return True
+
+    def _verify_table(self):
+        with sqlite3.connect(self.geopackage_file) as db:
+            cur = db.execute("""SELECT name FROM sqlite_master WHERE type='table' AND name=?""",
+                             (self.table_name,))
+            content = cur.fetchone()
+            if not content:
+                # Table doesn't exist _initialize_gpkg will create a new one.
+                return False
+            return True
+
+    def _verify_gpkg_contents(self):
+        with sqlite3.connect(self.geopackage_file) as db:
+            try:
+                cur = db.execute("""SELECT table_name, data_type, srs_id FROM gpkg_contents WHERE table_name = ?"""
+                                 , (self.table_name,))
+            except sqlite3.OperationalError as oe:
+                if 'no such column' in str(oe):
```

Comment:
Is there a better way to do these checks? The message can change with other SQLite versions.